Add uTVL_KPI_UMA  as a Price Identifier (#204)
UMAprotocol/UMIPs: UMIPs/Add uTVL_KPI_UMA  as a Price Identifier.md

## HEADERS
| UMIP [#]     |                                                                                                                                  |
|------------|------------------------------------------------------------------------------------------------------------------------------------------|
| UMIP Title | [Add uTVL_KPI_UMA  as a Price Identifier
                                                                                                  |
| Authors    | [Hart Lambur hart@umaproject.org, Mhairi McAlpine mhairi@umaproject.org]
	       
| Status     | [Draft]

> | Created    | [4th March 2021] 

Reviewer: The formatting here is a little off. Suggest pulling from another UMIP

| [Link to Discourse   -  https://discourse.umaproject.org/t/add-utvl-kpi-uma-as-a-price-identifier/318]                                                                               
<br>
<br>
<br>

# SUMMARY 

> Key Performance Indicator(KPI) options are synthetic tokens that redeem to protocol tokens with the redemption value determined by performance against that indicator. One example of a KPI is Total Value Locked (TVL).  

Reviewer: ```suggestion
Key Performance Indicator (KPI) options are synthetic tokens that redeem to protocol tokens with the redemption value determined by performance against that indicator. One example of a KPI is Total Value Locked (TVL).  
```

> This UMIP enables the DVM to support price requests based on the TVL of UMA

Reviewer: ```suggestion
This UMIP enables the DVM to support price requests based on the TVL of UMA.
```

> A synthetic option is minted against a base collateral, in this case UMA, which expires at 00.00(UTC) June 30th 2021.

Reviewer: I don't think this would be specific to a certain expiry date, would it?

Options are redeemed on the basis of TVL/10^9, with a floor of 0.1 and a ceiling of 2.

The value locked is calculated using the methods included in previous UMIPs which have price identifies associated with collateral types where available, collateral types which do not have an associated Price Identifier, or which are not yet approved will follow these presidents where available. 

The dollar value of each of the collateral types is then summed to provide the total value locked.

  

# MOTIVATION

> The primary motivation for the development of KPI options is to allow protocols to incentivise Defi users to assist them to reach the protocol's identified targets by leveraging their community resources by sharing their value with their community members.  

Reviewer: ```suggestion
The primary motivation for the development of KPI options is to allow protocols to incentivize DeFi users to assist them to reach the protocol's identified targets by leveraging their community resources by sharing their value with their community members.  
```

> Total Value Locked(TVL) is a frequently quoted key performance indicator and one which has a level of prominence in key Defi dashboards as an indicator of the health of a protocol.  Within UMA, TVL plays the additional function of being the lower floor for the value of 0,5 $UMA, due to the design of UMA and the COC>POC inequality.

Reviewer: ```suggestion
Total Value Locked(TVL) is a frequently quoted key performance indicator and one which has a level of prominence in key DeFi dashboards as an indicator of the health of a protocol.  Within UMA, TVL plays the additional function of being the lower floor for the value of 0.5 $UMA's market cap, due to the design of UMA and the COC>POC inequality.
```

1. What are the financial positions enabled by creating this synthetic that do not already exist?

  - The DVM does not currently support any form of KPI options.  This synthetic token will allow the creation of tokens which expire to a set rate of the collateral asset tokens based on a pre-identified bounded ratio as determined by the TVL of the protocol at the time of expiry.

2. Please provide an example of a person interacting with a contract that uses this price identifier. 

 - A protocol may choose to leverage its community and/or its reputation by minting TVL Options for its token which can be redeemed to a token amount as determined by the TVL of the protocol at the expiry point.

>  - A protocol community member, tokenholder, voter or proximal Defi prtocol participant may be gifted a TVL option by a protocol as an incentive to build the TVL of the protocol within the option timeframe and redeem at expiry.

Reviewer: ```suggestion
 - A protocol community member, tokenholder, voter or proximal DeGi protocol participant may be gifted a TVL option by a protocol as an incentive to build the TVL of the protocol within the option timeframe and redeem at expiry.
```

 - Any user may purchase a TVL Option for a protocol that they believe has the potential for growth in TVL prior to expiry.   

3. The current TVL of UMA is approximately $83m as at 21.30(UTC) 5th March 2021.

<br> 

# MARKETS & DATA SOURCES

There are a variety of assets approved as Collateral within UMA's priceless contracts.
A list of these assets is [available on our docs site](https://docs.umaproject.org/uma-tokenholders/approved-collateral-currencies).  To maintain consistency with existing price identifier UMIPs, it is suggested that different markets are queried for different collateral types. These are grouped below. 

 - [wEth] - the wrapped native token of the Ethereum Network (henceforth referred to as ETH)
 - [renBTC, wBTC] - wrapped ERC20 versions of Bitcoin (henceforth referred to as BTC)
 - [DAI, USDC, USDT, rDai]  - ERC20 dollar stable coins (henceforth referred to as USD)

> Additionally there are a number of liquidity tokens that are accepted as collateral, as well as "native" ERC20 tokens.

Reviewer: ```suggestion
Additionally there are a number of LP tokens that are accepted as collateral, as well as "native" ERC20 tokens.
```

Note that the price identifier of UMIP 7 is depreciated, see rationale for further discussion

	3. USD 	- assumed to be exactly 1USD. 
Should other non-algorithmic USD stablecoins be added to the accepted collateral currency whitelist prior to expiry, these should also be similarly grouped and assumed to be worth exactly 1USD
Note - See rationale for further discussion

	4. Liquidity Provider Tokens.  	- follow specifications as per UMIP 59 unless there is a known reason for devience)
https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-59.md
Should other Uniswap liquidity tokens be added to the accepted collateral currency whitelist prior to expiry, these should also be similarly calculated.
Note - see rationale for further discussion

	5. Native ERC20 Tokens 
		- where a price identifier UMIP exists that defines a dollar value, this should be used.
		- where no price identifier exists,the highest volume USD* markets should be used.

**Note - see rationale for further discussion**

> 2.  Which specific pairs should be queried from each market?

Reviewer: Suggest grouping all of these sections as the answers are all the same

	4. Liquidity Provider Tokens 	- see UMIP 59
	5. Native ERC20 tokens - see relevant PI UMIPs, where no UMIP exists, provided price update tbd

5. Provide recommended endpoints to query for historical prices from each market listed. 

	1. ETH 	- see UMIP 6
	2. BTC 	- see UMIP 7
	3. USD 	 - n/a - historical price is always $1
	4. Liquidity Provider tokens - see UMIP 59
	5. Native ERC20 tokens - see relevant PI UMIPs, where no UMIP exists endpoint tbd.

6.  Do these sources allow for querying up to 74 hours of historical data? 

  	- See relevant UMIPs
 	

7.  How often is the provided price updated?

   	- Where updated prices are required, relevant UMIPs refer to the frequency
	

8. Is an API key required to query these sources? 
	
	 - See Relevant UMIPs
 	

9. Is there a cost associated with usage? 

	- See relevant UMIPS
	
10. If there is a free tier available, how many queries does it allow for?

    - The cost impact of this PI would be negligible.

11.  What would be the cost of sending 15,000 queries?

 - There is no need to sent 15, 000 queries for this price identifier as it does not require bots. 

<br>

# PRICE FEED IMPLEMENTATION

> As there is no requirement for ongoing monitoring through liquidation or dispute bots, a price feed is not required.  The only requirement is a query of the UMA TVL statistic at the timestamp 00.00(UTC) on 30th June according to the data and markets as defined above.

Reviewer 2: So Risk Labs is the only allowed sponsor? Is this price id a one use price id (based on synth expiry time being hardcoded into the price id spec itself and the fact that proper collateralization up to 2 UMA price is a given by Risk Labs)?

Author: I think it is a one-use UMIP, if we were to create new KPI options we would probably put forward a new PI

Reviewer: We should also list this [script](https://github.com/UMAprotocol/protocol/blob/master/packages/merkle-distributor/kpi-options-helpers/calculate-uma-tvl.ts), which calculates UMA TVL.

Please be aware that the value of any UMA synthetic token is the value of the price identifier in units of the collateral currency used. If a contract’s price identifier returns 1, and is collateralized in renBTC, each synthetic will be worth 1 renBTC. In most cases, the value of your quote currency and intended collateral currency should be equal.

- The collateral redemption is designed to be tied to the value of the TVL of the protocol by design.

**4. Intended Collateral Currency** 
 
 - UMA

Does the value of this collateral currency match the standalone value of the listed quote currency? 

 - No, this is a design feature.

Is your collateral currency already approved to be used by UMA financial contracts? If no, submit a UMIP to have the desired collateral currency approved for use. 

 - YES 
 - UMA was approved as a collateral currency in[UMIP 56](https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-56.md) 

**5. Collateral Decimals** 

 - 18

**6. Rounding** - Round to 2 decimal places.

 - If the price is 1.025, then this would be rounded up to 1.03. If the price is 1.0249999, then this would be rounded down to 1.02.

>  - if the value returned is less than 0.05 round up to 0.1 to provide a floor price.

Reviewer: ```suggestion
 - if the value returned is less than 0.01 round up to 0.1 to provide a floor price.
```

Reviewer 3: What if the value returned is between 0.01 and 0.1? I understood that everything below 0.1 should be rounded up to 0.1 for floor price

It is assumed that for this purpose, 1 renBTC=1wBTC=1BTC.  While that assumption may be somewhat inaccurate, it has minimal impact on the overall calculation,  cuts down on processing complexity, and works on a “common sense” attitude that wrapped bitcoin on the Eth chain are equivalent in value to native bitcoin.  
There are other forms of wrapped BTC which are not approved as collateral within the UMA protocol.  To maintain consistency it is suggested that this assumption is made for any other similar non-algorithmic wrapped Bitcoin.  

3. Four non-algorithmic dollar stablecoins (DAI, USDC, USDT, rDai]) are approved as collateral within the UMA Protocol. 
It is assumed for this purpose that 1DAI=1USDC=1USDT=1rDAI=$1.  While this assumption may be somewhat inaccurate, it has minimal impact on the overall calculation, cuts down on processing complexity and works on a “common sense” attitude that a dollar stablecoin is worth exactly $1.
There are other forms of dollar stablecoins which are not approved as collateral within the UMA protocol.  To maintain consistency it is suggested that this assumption is made for any other non-algorithmic stablecoin. 

> 4. FiveLiquidity Tokens (wBTC-Eth, USDC-Eth, UNI-Eth, UMA-Eth, bwBTC/ETH_SLP) are approved as collateral within the UMA protocol. 

Reviewer 2: I think bwBTC/ETH_SLP should be moved to the native ERC20 tokens because of the extra step of pricePerFullShare. Using UMIP59 to calculate bwBTC/ETH_SLP/USD would return a lower price (higher if the badger sett gets drained) than the real one.

Author: See comment above.

5.There are a variety of price identifiers for native ERC20 tokens.  Calculation of the dollar value should follow the relevant UMIP where it exists, and where there is no Price Identifier UMIP, the highest volume [ERC20]/USD* market should be used.

 - There is no need for price processing.  This is a snapshot based on a particular time, however it may be useful for TVL Options holders to have oversight of the ongoing TVL and consequently the value of their options on an ongoing basis.  There are currently two dashboards that track the TVL of UMA - 
	- [SimpleID](https://monitor.simpleid.xyz/uma)
 	- [Yuen](https://docs.google.com/spreadsheets/d/e/2PACX-1vSEMURxiVQuu6jSDp2zmI7kdKKaJjgmhWNiVjwStyJekDx9hWgclKzm_yv9iyj82IRP4d9dZ8rgvCCB/pubhtml) 

 - These currently monitor different contracts and calculate the value slightly differently.

 - The use of these dashboards was considered but rejected due to the potential for manipulation and the differences between their calculations of the values of the collateral assets and the value as determined by the relevant UMIPs.

<br>

# IMPLEMENTATION

 - The contracts to be tracked include 
	- Any EMP contract
 	- Any Perpetual contract

> 	- Any perpetual pool party contract

Reviewer: Are we including perpetual pool party contracts? I think this is a bit of an open question

 - The total value locked is the value of the collateral locked in the relevant contracts.
 - The dollar value of each of the contracts should be calculated using the UMIPs and guidence in the Markets and Data section.  
 - These should then be summed to obtain the total value locked (TVL) measured in dollars.
 - The TVL as measured in dollars should then be divided by 10^9

1. **What prices should be queried for and from which markets?**

 - The dollar value of each of the contracts should be calculated using the UMIPs and guidence in the Markets and Data section. 
 - Where there is no price identifier UMIP, the price should be queried from the highest volume USD* market

2. **Pricing interval**

    - This will vary for each value however the relevant UMIPs provide a guide to the pricing interval for each

3. **Input processing**

>     - From the list of ExpiringMultiPartyCreator (available at https://github.com/UMAprotocol/protocol/blob/master/packages/affiliates/payouts/devmining-status.json), query all EMP addresses from createExpiringMultiParty() and call pfc() on all of them, repeat same steps for PerpetualCreator and just check the createPerpetual()  event, to identify the amount of each collateral type locked in UMA contracts. 

Author: @daywiss could you comment?
@smb2796 suggested this method and I'm not sure how practical the proposed alternative above is.

>     - From the list of ExpiringMultiPartyCreator (available at https://github.com/UMAprotocol/protocol/blob/master/packages/affiliates/payouts/devmining-status.json), query all EMP addresses from createExpiringMultiParty() and call pfc() on all of them, repeat same steps for PerpetualCreator and just check the createPerpetual()  event, to identify the amount of each collateral type locked in UMA contracts. 

Reviewer: That is actually a list of ExpiringMultiParty's, not the Creator. Using this list is unnecessary since we are finding all EMP's from the creator.

 - The dollar value of each contract should then be calculated using the details supplied in the Markets and Data section referencing the relevant UMIPs

 - These should then be summed to obtain the total value locked (TVL) measured in dollars.

4. **Result processing** 

>     - Divide TVL by 10^9 and apply the food and ceiling price rounding.

Author: oops!

<br>

> # Security considerations

Reviewer 3: What are implications of rogue liquidation attacks on the KPI options program? E.g. if at some point TVL reaches $2bn, and someone tries liquidating part of sponsors' debt position at uTVL_KPI_UMA>2, which is disputed and DVM returns price of 2 - do I understand correctly, that the sponsor's debt position would still be settled at 2? Now, if at expiration TVL has fallen (e.g. to $1bn), the sponsor would not be able to get out excess collateral at price of 1 because part of debt already has been settled for the price of 2.
Sure, there is no economic gain for the liquidator (just losses), but is there a risk that part of UMA collateral could get stuck in the contract?

Author: It seems unlikely given the lack of economic gain, but in such a case we could settle the dispute about which is the correct TVL using the DVM

1. Where could manipulation occur?

>  - Negligible opportunities for manipulation.

Reviewer 4: Is the TVL calculated as a TWAP at all? Otherwise I could see a pretty simple manipulation where a KPI holder deposits a ton of collateral into a contract last second before the snapshot and then withdraws. @chrismaree and I have been discussing this offline.

Reviewer: I'm in favor of this, but think the TWAP would need to be considerably long to prevent this manipulation. If we did something like 2 hours, is it really that much more painful to just keep the capital in the contracts for 2-hours rather than a few seconds? Something like a week might make sense.

Author: I think this is a good point, but unsure how to resolve given the variety of markets.

Chris suggested that we could use coingecko for individual ERC20s but there is the problem of LP tokens which dont have a dollar price feed.

Reviewer 2: Wouldn't applying full TWAP to the final price (aka to every underlying price) be cumbersome for sources that don't natively provide it?

Can averaging snapshots over the suggested period alleviate this (although exploiter could easily jump in and out)?

What about using random snapshots over the period, where the random timestamps are decided after expiry via any public shareable source of randomness (not sure how good the entropy would be, but something like using the hash of the first block after expiry as seed)? So everyone knows how to calculate the random TS beforehand but nobody knows what the chosen TS will be up until expiry.

## HEADERS
| UMIP [#]     |                                                                                                                                  |
|------------|------------------------------------------------------------------------------------------------------------------------------------------|

> | UMIP Title | [Key Performance Indicator; Total Value Locked UMA Price Identifier                                                                                                  |

Reviewer: To follow the existing convention, the title should just be: "Add uTVL_KPI_UMA as a price identifier".

A list of these assets is [available on our docs site](https://docs.umaproject.org/uma-tokenholders/approved-collateral-currencies).  To maintain consistency with existing price identifier UMIPs, it is suggested that different markets are queried for different collateral types. These are grouped below. 

 - [wEth] - the wrapped native token of the Ethereum Network (henceforth referred to as ETH)
 - [renBTC, wBTC] - wrapped ERC20 versions of Bitcoin (henceforth referred to as BTC)
 - [DAI, USDC, USDT, rDai]  - ERC20 dollar stable coins (henceforth referred to as USD)
 - [wBTC-Eth, USDC-Eth, UNI-Eth, UMA-Eth] -UNI Liquidity Tokens (henceforth referred to as uLQ)
 - [bwBTC/ETH SLP] Sushi Liquidity Tokens (henceforth referred to a sLQ)

Additionally there are a number of “native” ERC20 tokens which are also accepted as collateral, namely
 - bBadger, 
 - PERL, 
 - DSD, 
 - renDoge, 
 - OCEAN, 
 - YAM, 
 - Aave, 
 - Link, 
 - SNX, 
 - UMA, 
 - UNI

It is proposed that these are treated in the above groups for the purposes of determining markets and data sources.  For discussion on this see #Rationale 

>  **Required questions**

Reviewer: This is all good information, but I wonder if it is overcomplicating things. Perhaps we should just define some heuristics around how to choose the pricing source for each collateral. It is likely that there will be additional collateral types that are not covered by this list, and defining rules rather than explicit instructions for each would cover these.

What do you think about?
- For collateral types that have a Collateral/USD price identifier already, use the instructions from that price id.
- For types that do not, use the price from the highest 24 hour volume exchange.

Author: This doesnt work for Liquidity tokens,  but yes....this can be simplified a lot.

Author: I've stuck with the Weth, BTC and USD stablecoin definitions and grouped the liquidity tokens together to follow UMIP 59.  The rest can be either UMIPs or derived from highest volume dollar markets. I've also added in the rationale notes that where there is no dollar market, the highest volume eth market should be used, then dollar value calculated from UMIP 6.

1. What markets should the price be queried from? It is recommended to have at least 3 markets.

	1. ETH 	- Kraken, Binance and Coinbase (as per UMIP 6)
https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-6.md)

	2. BTC 	-  Binance, Coinbase and Bitstamp (as per UMIP 7)
https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-7.md
Should other non-algorithmic derivations of BTC to be added to the accepted collateral currency whitelist prior to expiry, these should also be similarly grouped and  follow the calculation methods detailed in UMIP 7, see rationale for further details.
Note that the price identifier of UMIP 7 is depreciated, see rationale for further discussion

	3. USD 	- assumed to be exactly 1USD. 
Should other non-algorithmic USD stablecoins be added to the accepted collateral currency whitelist prior to expiry, these should also be similarly grouped and assumed to be worth exactly 1USD
Note - See rationale for further discussion

> 	4. uLQ 	- Uniswap (calculations performed as per UMIP 59)

Reviewer: What is uLQ and sLQ?

Author: Uniswap Liquidity tokens and Sushiswap liquidity tokens 

I'm suggesting that any other Uniswap or Sushiswap LQ tokens that are approved as collateral types follow these conventions unless there is good reason not to.

> https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-59.md

Reviewer 2: This UMIP doesn't define LP/USD, only USD/LP. LP/USD is the result of the previous to last step of the implementation, so I'd add the step number to this reference.

It also applies to 7. bBadger - Sushiswap

Author: I dont think its relevant for markets and data (as the same sources apply), however have added it into the rationale to be explicit about where the calculation should come from.

Should other Uniswap liquidity tokens be added to the accepted collateral currency whitelist prior to expiry, these should also be similarly calculated.
Note - see rationale for further discussion

	5. sLQ 	- Sushiswap (calculations performed as per UMIP 39)

> https://github.com/UMAprotocol/UMIPs/blob/master/UMIPs/umip-39.md

Reviewer 2: I don't think this UMIP is the best example for SLPs calculation because the SLPs in UMIP 39 are wrapped through a  Badger Sett vault so the calculation takes into account the pricePerFullShare of the vault.

 I think Uniswap and Sushiswap calculations can be unified, they are the same implementation wise.

Author: Have combined uniswap and sushiswap calculations to follow UMIP 59

Author: It could go in as a native ERC20 token section to use UMIP39, or grouped together with the other LPs, as per UMIP 59. 

Which do you think is better?

Reviewer 2: I think it should go as native linked with UMIP39, because it's slightly diff than a standard LP token.

> Should other Uniswap liquidity tokens be added to the accepted collateral currency whitelist prior to expiry, these should also be similarly calculated.

Reviewer 2: Should say Sushiswap

3. Four non-algorithmic dollar stablecoins (DAI, USDC, USDT, rDai]) are approved as collateral within the UMA Protocol. 
It is assumed for this purpose that 1DAI=1USDC=1USDT=1rDAI=$1.  While this assumption may be somewhat inaccurate, it has minimal impact on the overall calculation, cuts down on processing complexity and works on a “common sense” attitude that a dollar stablecoin is worth exactly $1.
There are other forms of dollar stablecoins which are not approved as collateral within the UMA protocol.  To maintain consistency it is suggested that this assumption is made for any other non-algorithmic stablecoin. 

4. Four Uniswap Liquidity Tokens (wBTC-Eth, USDC-Eth, UNI-Eth, UMA-Eth) are approved as collateral within the UMA protocol. 
It is recommended that any additional proposal to add Uniswap liquidity tokens as approved collateral currencies are evaluated against UMIP 59 and follow those market and data sources unless there is good reason for deviance

> 5. One Sushisaw Liquidity Token ( [bwBTC/ETH SLP) is approved as collateral in the UMA Protocol. 

Reviewer 5: Sushisaw -> sushiswap

6. RenDoge is approved as a collateral currency in UMA, however there is no associated price identifier UMIP.  
It is recommended that for this purpose, 1 RenDoge = 1 Doge.  While that assumption may be somewhat inaccurate, it has minimal impact on the overall calculation,  cuts down on processing complexity, and works on a “common sense” attitude that wrapped Doge on the Eth chain are equivalent in value to native doge. The most common doge pair is Doge/USDT, the three markets with greatest volume as reported by Coingecko were chosen.

 - There is no need for price processing.  This is a snapshot based on a particular time, however it may be useful for TVL Options holders to have oversight of the ongoing TVL and consequently the value of their options on an ongoing basis.  There are currently two dashboards that track the TVL of UMA - 
	 - [SimpleID](https://monitor.simpleid.xyz/uma)
 	- [Yuen](https://docs.google.com/spreadsheets/d/e/2PACX-1vSEMURxiVQuu6jSDp2zmI7kdKKaJjgmhWNiVjwStyJekDx9hWgclKzm_yv9iyj82IRP4d9dZ8rgvCCB/pubhtml) 

 - These currently monitor different contracts and calculate the value slightly differently.

>  - The use of these dashboards was considered but rejected due to the potential for manipulation and the differences between their calculations of the values of the collateral assets and the value as determined by the relevant UMIPs.

Reviewer 5: What methodologies do the relevant UMIPs use to calculate the value?

Author: There is a bit of variety in markets and data sources.

<br>

# IMPLEMENTATION

>  - The dollar value of each of the contracts should be calculated using the UMIPs and guidence in the Markets and Data section.  

Reviewer: I think we need to cover which contracts need to be used in this calculation within this section. For right now, I believe the rules would be:

1. Any EMP contract
2. Any Perpetual contract
3. Any perpetual pool party contract

Author: Whats a perpetual pool party contract?  Is that the optimistic oracle?

Author: Have added that in.

>  - The dollar value of each of the contracts should be calculated using the UMIPs and guidence in the Markets and Data section.  

Reviewer: Let's also be explicit about the fact that we are using the total value locked of the collateral currency in each contract.

 - These should then be summed to obtain the total value locked (TVL) measured in dollars.

>  - The TVL as measured in dollars should then be divided by 10^8

Reviewer: Shouldn't this be divided by 10^9, not 10^8? I think it'd be clearer to just say that this needs to be divided by 1 billion

Author: Oh, yeah, thats changed from the original KPI options paper.

Will update, I've referred to 100 million in a couple of places.
